### PR TITLE
[SPARK-40465][SQL] Refactor Decimal so as we can use other underlying implementation

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
**The background**
Currently, Spark `Decimal` use the `java.math.BigDecimal` as the underlying implementation.
As we know, the computing and storage efficiency of Spark `Decimal`  is very low.
This PR let `Decimal` can supports other underlying implementation.

**The design**
The refactoring design of Spark `Decimal` is mainly based on the proxy mode - that is, to create the abstraction class `DecimalOperation`, to lower the current implementation of Spark `Decimal` based on `java.math.BigDecimal` to `JDKDecimalOperation`. The API of Spark `Decimal` remains unchanged and proxies the corresponding interface methods of `DecimalOperation`. The whole design is shown in the figure below.
UML 图

Whether Spark `Decimal` will proxy `JDKDecimalOperation` or other subclasses of `DecimalOperation` depends on the configuration parameter `spark.sql.decimal.implementation`. The default value of `spark.sql.decimal.implementation` is `JDKBigDecimal`, that is, `Decimal` will proxy `JDKDecimalOperation`. `spark.sql.decimal.implementation` can also be set as other possible optimization implementations.

**Benchmark**
Before this refactor, the decimal benchmark show below.
```
[info] running (fork) org.apache.spark.sql.DecimalBenchmark 
[info] Running benchmark: Decimal For bigDecimal and smallDecimal
[info]   Running case: Operator +
[info]   Stopped after 2 iterations, 30342 ms
[info]   Running case: Operator -
[info]   Stopped after 2 iterations, 33470 ms
[info]   Running case: Operator *
[info]   Stopped after 2 iterations, 36113 ms
[info]   Running case: Operator /
[info]   Stopped after 2 iterations, 62917 ms
[info]   Running case: Operator %
[info]   Stopped after 2 iterations, 28986 ms
[info] Java HotSpot(TM) 64-Bit Server VM 1.8.0_311-b11 on Mac OS X 10.16
[info] Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
[info] Decimal For bigDecimal and smallDecimal:  Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] Operator +                                        15105          15171          94          8.9         112.5       1.0X
[info] Operator -                                        16134          16735         850          8.3         120.2       0.9X
[info] Operator *                                        17650          18057         576          7.6         131.5       0.9X
[info] Operator /                                        30421          31459        1468          4.4         226.7       0.5X
[info] Operator %                                        14342          14493         213          9.4         106.9       1.1X
[info] Running benchmark: Decimal For smallDecimal and bigDecimal
[info]   Running case: Operator +
[info]   Stopped after 2 iterations, 36101 ms
[info]   Running case: Operator -
[info]   Stopped after 2 iterations, 36972 ms
[info]   Running case: Operator *
[info]   Stopped after 2 iterations, 29374 ms
[info]   Running case: Operator /
[info]   Stopped after 2 iterations, 86810 ms
[info]   Running case: Operator %
[info]   Stopped after 2 iterations, 153669 ms
[info] Java HotSpot(TM) 64-Bit Server VM 1.8.0_311-b11 on Mac OS X 10.16
[info] Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
[info] Decimal For smallDecimal and bigDecimal:  Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] Operator +                                        17120          18051        1316          7.8         127.6       1.0X
[info] Operator -                                        17716          18486        1089          7.6         132.0       1.0X
[info] Operator *                                        14505          14687         258          9.3         108.1       1.2X
[info] Operator /                                        42314          43405        1543          3.2         315.3       0.4X
[info] Operator %                                        75702          76835        1601          1.8         564.0       0.2X
[info] Running benchmark: Decimal For smallDecimal and smallDecimal
[info]   Running case: Operator +
[info]   Stopped after 2 iterations, 5794 ms
[info]   Running case: Operator -
[info]   Stopped after 2 iterations, 5822 ms
[info]   Running case: Operator *
[info]   Stopped after 2 iterations, 31043 ms
[info]   Running case: Operator /
[info]   Stopped after 2 iterations, 66493 ms
[info]   Running case: Operator %
[info]   Stopped after 2 iterations, 56967 ms
[info] Java HotSpot(TM) 64-Bit Server VM 1.8.0_311-b11 on Mac OS X 10.16
[info] Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
[info] Decimal For smallDecimal and smallDecimal:  Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] -------------------------------------------------------------------------------------------------------------------------
[info] Operator +                                          2837           2897          84         47.3          21.1       1.0X
[info] Operator -                                          2897           2911          21         46.3          21.6       1.0X
[info] Operator *                                         14574          15522        1340          9.2         108.6       0.2X
[info] Operator /                                         32050          33247        1693          4.2         238.8       0.1X
[info] Operator %                                         28359          28484         177          4.7         211.3       0.1X
[info] Running benchmark: Decimal For bigDecimal and bigDecimal
[info]   Running case: Operator +
[info]   Stopped after 2 iterations, 4692 ms
[info]   Running case: Operator -
[info]   Stopped after 2 iterations, 4850 ms
[info]   Running case: Operator *
[info]   Stopped after 2 iterations, 31087 ms
[info]   Running case: Operator /
[info]   Stopped after 2 iterations, 64451 ms
[info]   Running case: Operator %
[info]   Stopped after 2 iterations, 58293 ms
[info] Java HotSpot(TM) 64-Bit Server VM 1.8.0_311-b11 on Mac OS X 10.16
[info] Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
[info] Decimal For bigDecimal and bigDecimal:    Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] Operator +                                         2332           2346          20         57.6          17.4       1.0X
[info] Operator -                                         2384           2425          59         56.3          17.8       1.0X
[info] Operator *                                        15188          15544         503          8.8         113.2       0.2X
[info] Operator /                                        31247          32226        1384          4.3         232.8       0.1X
[info] Operator %                                        28878          29147         380          4.6         215.2       0.1X
[success] Total time: 1369 s (22:49), completed 2022-9-21 19:42:19
```

After this refactor, the decimal benchmark show below.
```
================================================================================================
 bigDecimal and smallDecimal
 ================================================================================================

 Java HotSpot(TM) 64-Bit Server VM 1.8.0_311-b11 on Mac OS X 10.16
 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
 Decimal For bigDecimal and smallDecimal:  Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
 ------------------------------------------------------------------------------------------------------------------------
 Operator +                                        17969          18788        1159          7.5         133.9       1.0X
 Operator -                                        17755          22272        1981          7.6         132.3       1.0X
 Operator *                                        16503          17196         980          8.1         123.0       1.1X
 Operator /                                        33914          35547        2310          4.0         252.7       0.5X
 Operator %                                        15099          15873        1095          8.9         112.5       1.2X


 ================================================================================================
 smallDecimal and bigDecimal
 ================================================================================================

 Java HotSpot(TM) 64-Bit Server VM 1.8.0_311-b11 on Mac OS X 10.16
 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
 Decimal For smallDecimal and bigDecimal:  Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
 ------------------------------------------------------------------------------------------------------------------------
 Operator +                                        18722          18770          68          7.2         139.5       1.0X
 Operator -                                        17709          17775          95          7.6         131.9       1.1X
 Operator *                                        15323          15388          93          8.8         114.2       1.2X
 Operator /                                        40692          40700          11          3.3         303.2       0.5X
 Operator %                                        74394          77301         NaN          1.8         554.3       0.3X


 ================================================================================================
 smallDecimal and smallDecimal
 ================================================================================================

 Java HotSpot(TM) 64-Bit Server VM 1.8.0_311-b11 on Mac OS X 10.16
 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
 Decimal For smallDecimal and smallDecimal:  Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
 -------------------------------------------------------------------------------------------------------------------------
 Operator +                                          3732           3763          45         36.0          27.8       1.0X
 Operator -                                          3369           3376           9         39.8          25.1       1.1X
 Operator *                                         16184          16779         841          8.3         120.6       0.2X
 Operator /                                         33275          33616         483          4.0         247.9       0.1X
 Operator %                                         28803          29009         291          4.7         214.6       0.1X


 ================================================================================================
 bigDecimal and bigDecimal
 ================================================================================================

 Java HotSpot(TM) 64-Bit Server VM 1.8.0_311-b11 on Mac OS X 10.16
 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
 Decimal For bigDecimal and bigDecimal:    Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
 ------------------------------------------------------------------------------------------------------------------------
 Operator +                                         3819           3959         197         35.1          28.5       1.0X
 Operator -                                         3379           3450         100         39.7          25.2       1.1X
 Operator *                                        16278          16808         749          8.2         121.3       0.2X
 Operator /                                        33721          34187         659          4.0         251.2       0.1X
 Operator %                                        28911          29671        1075          4.6         215.4       0.1X
```

We can see the proxy mode have the overhead to create `DecimalOperation` instances.

### Why are the changes needed?
Refactor `Decimal` so as we can use other underlying implementation.

### Does this PR introduce _any_ user-facing change?
'No'.
The interface of `Decimal` is not changed.

### How was this patch tested?
N/A

### CODE REVIEW

**Comment by baibaichen** [comment]
@beliefer thanks for bring this up!

cc @cloud-fan would you pelease review this PR.

**Comment by beliefer** [comment]
This PR can ensure that the existing underlying implementation of `Decimal` will not be changed, and it also provides an interface for us to develop new implementations through iteration. With this interface, no matter whether we can enter the community in the future, our development and maintenance costs will be greatly reduced.

**Comment by cloud-fan** [comment]
The abstraction makes sense to me. The new `DecimalOperation` layer adds an indirection and may hurt performance, though I think the overhead is minor as Java decimal operation is already slow. Can we add a microbenchmark for `Decimal`? You can follow `HashBenchmark`.

**Comment by beliefer** [comment]
> The abstraction makes sense to me. The new `DecimalOperation` layer adds an indirection and may hurt performance, though I think the overhead is minor as Java decimal operation is already slow. Can we add a microbenchmark for `Decimal`? You can follow `HashBenchmark`.

I followed `HashBenchmark` and create a new benchmark `DecimalBenchmark`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala [hunk 1]
```diff
@@ -28,52 +28,35 @@ import org.apache.spark.sql.internal.SQLConf
...
 @Unstable
-final class Decimal extends Ordered[Decimal] with Serializable {
+final class Decimal(initEnabled: Boolean = true) extends Ordered[Decimal] with Serializable {
   import org.apache.spark.sql.types.Decimal._
 
-  private var decimalVal: BigDecimal = null
-  private var longVal: Long = 0L
-  private var _precision: Int = 1
-  private var _scale: Int = 0
+  private var decimalOperation: DecimalOperation[_] = null
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala:37]: does it mean `decimalOperation` can be null sometimes? This seems fragile.

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala:37]: decimalOperation is not null in default. To reduce the overhead of operations (e.g. +), let it is null.
I tested the benchmark, if give a default `DecimalOperation` value, these math operations (e.g. +) have 5x performance overhead.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/DecimalBenchmark.scala [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+/**
+ * Benchmark for the previous Decimal (without proxy mode) vs refactored Decimal (proxy mode).
+ * To run this benchmark:
+ * {{{
+ *   1. without sbt:
+ *      bin/spark-submit --class <this class> --jars <spark core test jar> <spark catalyst test jar>
+ *   2. build/sbt "catalyst/Test/runMain <this class>"
+ *   3. generate result:
+ *      SPARK_GENERATE_BENCHMARK_FILES=1 build/sbt "catalyst/Test/runMain <this class>"
+ *      Results will be written to "benchmarks/DecimalBenchmark-results.txt".
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/DecimalBenchmark.scala:34]: can you run it to generate result and commit the result?

Comment (beliefer) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/DecimalBenchmark.scala:34]: OK

**Comment by beliefer** [comment]
ping @cloud-fan would you pelease review this PR.

**Pull request closed without merging**